sound/soc/sof/pcm.c

```diff
@@ -116,6 +116,40 @@ static int sof_pcm_dsp_pcm_free(struct snd_pcm_substream *substream,
 	return ret;
 }
 
+static int sof_pcm_setup_connected_widgets(struct snd_sof_dev *sdev,
+					   struct snd_soc_pcm_runtime *rtd,
+					   struct snd_sof_pcm *spcm, int dir)
+{
+	struct snd_soc_dai *dai;
+	int ret, j;
+
+	/* query DAPM for list of connected widgets and set them up */
+	for_each_rtd_cpu_dais(rtd, j, dai) {
+		struct snd_soc_dapm_widget_list *list;
+
+		ret = snd_soc_dapm_dai_get_connected_widgets(dai, dir, &list,
+							     dpcm_end_walk_at_be);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: dai %s has no valid %s path\n", dai->name,
+				dir == SNDRV_PCM_STREAM_PLAYBACK ? "playback" : "capture");
+			return ret;
+		}
+
+		spcm->stream[dir].list = list;
+
+		ret = sof_widget_list_setup(sdev, spcm, dir);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed widget list set up for pcm %d dir %d\n",
+				spcm->pcm.pcm_id, dir);
+			spcm->stream[dir].list = NULL;
+			snd_soc_dapm_dai_free_widgets(&list);
```

Comment:
I also struggled with the "NEW" and "FREE" verbs used in the SOF IPC interface, and when these are used in kernel code. E.g. sof_widget_free() sends a FREE IPC to the firmware, But it does not free the "struct snd_sof_widget *swidget" in the sense "free" is commonly used in kernel. This is especially hard to follow when more conventional kernel functions are used in the same functions. E.g. snd_soc_dapm_dai_free_widgets() does free resources with a kfree() in the end.

But, but ,using terms that are not aligned with the SOF IPC command names, would lead to other kind of confusion, so not sure how to completely avoid the namespace collisions. And resources are freed in SOF, so it's not like it's complete misuse of the terms either.